notifyme-sdk-backend/notifyme-sdk-backend-ws/src/main/java/ch/ubique/notifyme/sdk/backend/ws/insertmanager/InsertManager.java

```diff
@@ -0,0 +1,118 @@
+package ch.ubique.notifyme.sdk.backend.ws.insertmanager;
+
+import ch.ubique.notifyme.sdk.backend.data.NotifyMeDataServiceV3;
+import ch.ubique.notifyme.sdk.backend.model.UserUploadPayloadOuterClass.UploadVenueInfo;
+import ch.ubique.notifyme.sdk.backend.ws.insertmanager.insertfilters.UploadInsertionFilter;
+import ch.ubique.notifyme.sdk.backend.ws.semver.Version;
+import ch.ubique.notifyme.sdk.backend.ws.util.CryptoWrapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.List;
+
+public class InsertManager {
+
+    private static final Logger logger = LoggerFactory.getLogger(InsertManager.class);
+
+    private final List<UploadInsertionFilter> filterList = new ArrayList<>();
+
+    private final CryptoWrapper cryptoWrapper;
+    private final NotifyMeDataServiceV3 notifyMeDataServiceV3;
+
+    public InsertManager(CryptoWrapper cryptoWrapper, NotifyMeDataServiceV3 notifyMeDataServiceV3) {
+        this.cryptoWrapper = cryptoWrapper;
+        this.notifyMeDataServiceV3 = notifyMeDataServiceV3;
+    }
+
+    /**
+     * Adds a filter to the list of filters to be applied to a {@link UploadVenueInfo} object before inserting it
+     * into the database as a {@link ch.ubique.notifyme.sdk.backend.model.tracekey.v3.TraceKey}.
+     * @param filter to be added to the filter list. The filter method must return a filtered list of VenueInfo's
+     */
+    public void addFilter(UploadInsertionFilter filter) {
+        this.filterList.add(filter);
+    }
+
+    /**
+     * Applies the stored list of filters, transforms any remaining venueInfo's into trace keys, and stores them to the
+     * database.
+     * @param uploadVenueInfoList List of UploadVenueInfo objects to be stored to the database
+     * @param header User-Agent header as included in the user-upload request
+     * @param principal the authorization context which belongs to the uploaded keys. This will usually be a JWT token.
+     * @param now Current timestamp
+     * @throws InsertException
+     */
+    public void insertIntoDatabase(
+            List<UploadVenueInfo> uploadVenueInfoList,
+            String header,
+            Object principal,
+            LocalDateTime now
+    ) throws InsertException {
+        if (uploadVenueInfoList != null && !uploadVenueInfoList.isEmpty()) {
```

Comment:
should we not throw an error if the uploadVenueInfoList is empty? an empty list is not a valid parameter for this request

Reply:
Resolved: See [https://github.com/SwissCovid/swisscovid-cn-backend/commit/9892dcf240971608d2b338b9867b79760f3461b0](https://github.com/SwissCovid/swisscovid-cn-backend/commit/9892dcf240971608d2b338b9867b79760f3461b0)
Note that the entire repository has been moved to [SwissCovid/swisscovid-cn-backend](https://github.com/SwissCovid/swisscovid-cn-backend) and future work will continue there.